src/FakeItEasy.Analyzer/NonVirtualSetupAnalyzer.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace FakeItEasy.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class NonVirtualSetupAnalyzer:DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = CreateSupportedDiagnostics();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeCall, SyntaxKind.InvocationExpression);
+        }
+
+        private static void AnalyzeCall(SyntaxNodeAnalysisContext context)
+        {
+            var invocationExpression = (InvocationExpressionSyntax)context.Node;
+
+            if (IsInterface(context, invocationExpression))
+                return;
+
+            if (NotSetupCall(invocationExpression))
+                return;
+
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(context.Node);
+
+            if (!symbolInfo.Symbol.IsVirtual)
+            {
+                var location = invocationExpression.GetLocation();
+
+                Diagnostic diagnostic = Diagnostic.Create(DiagnosticDefinitions.NonVirtualSetup, location, symbolInfo.Symbol.Name);
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static bool NotSetupCall(InvocationExpressionSyntax invocationExpression)
+        {
+            var isSetupInvocation = invocationExpression
+                .Ancestors()
+                .OfType<InvocationExpressionSyntax>()
+                .Any(x => x.Expression.ToString() == "A.CallTo");

```

Comment:
This doesn't handle calls made via a `using static FakeItEasy.A`. That is:
```c#
using static FakeItEasy.A;

…
        var foo = Fake<Foo>();
        CallTo(() => foo.Bar());
```

Will not trigger the diagnostic, even when `Bar` is not virtual. Perhaps following an approach closer to `UnusedReturnValueAnalyzer.AnalyzeCall` and `UnusedReturnValueAnalyzer.GetCalledMethodSymbol` would help.
This may help resolve other deficiencies as well. In particular, I think the whitelist of fully-qualified method names to examine (see `UnusedReturnValueAnalyzer.CreateDiagnosticsMap`) is quite helpful, in that it provides an obvious place to to reference to see what's being examined.

For example, even looking at that list now, I am reminded of ``A.CallToSet`1``, and ``FakeItEasy.Fake`.CallsTo`1``, which would also benefit from your diagnostic.

And the fully-qualifiedness of the names would help avoid false positives in the (extremely unlikely) event that the code being analyzed has another `A.CallTo`.